content/automate/ManageIQ/Cloud/VM/Retirement/StateMachines/Methods.class/__methods__/pre_retirement.rb

```diff
@@ -1,10 +1,35 @@
 #
 # Description: This method stops a cloud Instance
 #
+module ManageIQ
+  module Automate
+    module Cloud
+      module VM
+        module Retirement
+          module StateMachines
+            module Methods
+              class PreRetirement
+                def initialize(handle = $evm)
+                  @handle = handle
+                end
 
-vm = $evm.root['vm']
-if vm && vm.power_state == 'on'
-  ems = vm.ext_management_system
-  $evm.log('info', "Stopping Instance <#{vm.name}> in EMS <#{ems.try(:name)}>")
-  vm.stop if ems
+                def main
+                  vm = @handle.root['vm']
+                  if vm && vm.power_state == 'on'
+                    ems = vm.ext_management_system
```

Comment:
@tinaafitz if the ems is missing and the vm is powered on, should we be raising an exception and stopping the workflow or should we keep continuing. I remember we had some corner cases that we were trying to solve where a failed retirement is restarted. If there is no ems should we log the fact that its an orphaned vm.

Reply:
@mkanoor We should continue processing if the ems is nil and the VM will have the power state of "unknown" in that case